anatomy-of-locks.mdx

```diff
@@ -0,0 +1,187 @@
+---
+title: 'Anatomy of Table-Level Locks in PostgreSQL'
+description: 'This blog explains locking mechanisms in PostgreSQL, focusing on table-level locks that are required by Data Definition Language (DDL) operations.'
+image:
+  src: https://raw.githubusercontent.com/xataio/mdx-blog/main/images/anatomy-of-locks-og@2x.jpg
+  alt: Anatomy of Table-Level Locks in PostgreSQL
+author: Gulcin Yildirim Jelinek
+authorEmail: gulcin@xata.io
+date: 01-10-2025
+tags: ['postgres', 'postgresql', 'locks', 'ddl', 'dml', 'MVCC', 'pgroll']
+published: true
+slug: anatomy-of-locks
+canonicalUrl: https://pgroll.com/blog/anatomy-of-table-level-locks-in-postgresql
+ogImage: https://raw.githubusercontent.com/xataio/mdx-blog/main/images/anatomy-of-locks-og@2x.jpg
+---
+
+## Art of Locking or Unlocking?
+
+It is normal to think about database locks and go into physical-world lock analogies, and maybe order three books about the history of locks, Persian locks, and lock-picking techniques.
+Probably most of us learn by going deeper into the term “locking” to understand a concept in PostgreSQL that doesn’t really have much to do with physical-world locks, as physical locks are mostly about security.
```

Comment:
This intro reads a bit odd to me, maybe it could be:

```
It's common to think about database locks by drawing analogies to physical locks, which might even lead you to order books on the history of locks, Persian locks, and lock-picking techniques.

However, on deeper investigation of "locking" in PostgreSQL we come to realize it doesn't have much to do with physical locks at all, which are primarily about **security**.

Postgres locks though are all about **concurrency** and controlling which transaction can hold a lock while another transaction can do its thing, ideally without ever blocking each other.

But as we know, no world is ideal, whether it’s a door lock or an `AccessShareLock`.
```